### PR TITLE
RFC 0014: New Intrinsic Function to Convert Template Block to JSON String

### PR DESCRIPTION
**Pull request by mluk-aws**
### Language Enhancement Request For Comment

This is a request for comments about a new intrinsic function to convert template blocks to JSON strings. See #14 for
additional details.

---

## Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by jlhood — COMMENTED** [review]
Looks good. Just want to add syntax highlighting to examples. This is a good practice in general but for this feature in particular, it emphasize the syntax highlighting benefit of `Fn::ToJsonString`.

**Review by jlhood — APPROVED** [review]

File: RFCs/0014-ToJsonString.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                    - type: text
+                    x: 0
+                    y: 7
+                    width: 3
+                    height: 3
+                    properties:
+                        markdown: Hello world
+```
+
+# Motivation
```
Comment (MalikAtalla-AWS) [inline RFCs/0014-ToJsonString.md:87]: I think we decided to change our template to move the motivation section above examples

File: RFCs/0014-ToJsonString.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+A CloudFormation user may want to use JSON strings as input to a resource property. For example, `AWS::CloudWatch::Dashboard` requires a JSON string for the `DashboardBody` attribute. A user may even want to use a JSON string as input to attributes where only a general string is required, such as in `SecretString` of `AWS::SecretsManager::Secret`.
+
+Today, this can only be accomplished through the use of external tools to convert object bodies to JSON strings, or use workarounds such as YAML multiline syntax. Having a feature to automate this transformation can provide various benefits, such as improving development workflow, improving code readability, or being able to utilize JSON/YAML syntax highlighting in text editors.
```
Comment (MalikAtalla-AWS) [inline RFCs/0014-ToJsonString.md:91]: I wonder if we should explain a bit more here what the benefit of this feature in Yaml templates is. (Yaml multiline is a nice workaround, but IDEs and text editors treat the content as generic text, not yaml, so there is generally no syntax highlighting or validation)

File: RFCs/0014-ToJsonString.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+`Fn::ToJsonString` is an intrinsic function that takes in a template block as input and converts it into an escaped JSON string.
+
+* It will be restricted to only be used as the value of string-type resource properties.
+* Intrinsic functions (e.g. `Fn::If`, `Ref`) or pseudo parameters (e.g. `AWS::NoValue`) can be used within the input template block. The input template block will be processed by the intrinsic functions and pseudo parameters before it is converted to a string (i.e. the resolved value of the intrinsic function is converted to the output string, and not the intrinsic function itself).
+* Conversion will always retain the same order of key-value pairs such that the converted strings of the same input template block are guaranteed to not change.
+* The following pseudo parameters will be supported:
+    * `AWS::AccountId`, `AWS::Region`, `AWS::Partition`, `AWS::NoValue`
+
+# Limitation
```
Comment (benkehoe) [inline RFCs/0014-ToJsonString.md:105]: What is the cause of these limitations?

**Comment by benkehoe** [comment]
This is unrelated to the RFC itself, but I will say it in every forum available to me: no new resources should be allowed to ship that require string-containing-JSON, even if the underlying APIs require it (the resource handler should perform the conversion).

File: RFCs/0014-ToJsonString.md (hunk 2 above)
Comment (mluk-aws) [reply RFCs/0014-ToJsonString.md:91]: I can add a bit more explaining this limitation with YAML multiline.

File: RFCs/0014-ToJsonString.md (hunk 3 above)
Comment (mluk-aws) [reply RFCs/0014-ToJsonString.md:105]: These are short-term limitations due to implementation constraints.

**Comment by iann0036** [comment]
Per offline conversation, I would suggest that instead of making this an explicit action, we simply refer to the specification / registry type to determine the appropriate type at runtime. If the provided input is an object but the spec expects a string, simply perform the transformation automatically.

**Comment by benkehoe** [comment]
> determine the appropriate type at runtime

I'm skeptical of this. I would be fine if the type of these properties was "string containing JSON", but when it's just "string" I think coercing every string property value to string is removing the ability to lint and provide up-front errors, as you essentially eliminate *any* type errors for string-valued properties.

**Review by jlhood — APPROVED** [review]

File: RFCs/0014-ToJsonString.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+* YAML is a superset of JSON, so there are a few features in YAML that can not be converted to JSON due to their differences. For example, comments are supported in YAML but not JSON. This can potentially be a confusing developer experience if the developer is not aware of these limitations and have different expectations.
+* Unsupported pseudo parameters:
+    * `AWS::StackId`, `AWS::StackName`, `AWS::NotificationArns`, `AWS::URLSuffix`
```
Comment (benbridts) [inline RFCs/0014-ToJsonString.md:109]: I can see cases where these would be useful to have (eg. in the dashboard example: to create a text widget with a link back to the stack)

File: RFCs/0014-ToJsonString.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+* It will be restricted to only be used as the value of string-type resource properties.
+* Intrinsic functions (e.g. `Fn::If`, `Ref`) or pseudo parameters (e.g. `AWS::NoValue`) can be used within the input template block. The input template block will be processed by the intrinsic functions and pseudo parameters before it is converted to a string (i.e. the resolved value of the intrinsic function is converted to the output string, and not the intrinsic function itself).
+* Conversion will always retain the same order of key-value pairs such that the converted strings of the same input template block are guaranteed to not change.
+* The following pseudo parameters will be supported:
+    * `AWS::AccountId`, `AWS::Region`, `AWS::Partition`, `AWS::NoValue`
+
+# Limitation
+
+* YAML is a superset of JSON, so there are a few features in YAML that can not be converted to JSON due to their differences. For example, comments are supported in YAML but not JSON. This can potentially be a confusing developer experience if the developer is not aware of these limitations and have different expectations.
```
Comment (benbridts) [inline RFCs/0014-ToJsonString.md:107]: Can we be explicit about what happens with unsupported yaml features?
eg. I would like it to work even if there are comments inside the nested yaml.

Use case:

```yaml
MyDashboard:
    Type: AWS::CloudWatch::Dashboard
    Properties:
        DashboardBody:
            !ToJsonString:
                start: "-PT6H"  # 6 hours ago
[...]
```

Comment (mluk-aws) [reply RFCs/0014-ToJsonString.md:107]: Yes, the description here can be more clear. Your example with using comments in the yaml block should still be considered valid, but the json string it converts into will lose that comment (as the json data format does not support comments).

File: RFCs/0014-ToJsonString.md (hunk 4 above)
Comment (arthurboghossian) [reply RFCs/0014-ToJsonString.md:109]: The `AWS::StackId`, `AWS::StackName`, and `AWS::URLSuffix` pseudo parameters will be supported, but the `AWS::NotificationARNs` pseudo parameter will not be supported. Will update the RFC accordingly.

**Review by arthurboghossian — DISMISSED** [review]

File: RFCs/0014-ToJsonString.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+* Conversion will always retain the same order of key-value pairs such that the converted strings of the same input template block are guaranteed to not change.
+* The following pseudo parameters will be supported:
+    * `AWS::AccountId`, `AWS::Region`, `AWS::Partition`, `AWS::NoValue`, `AWS::StackId`, `AWS::StackName`, `AWS::URLSuffix`
+
+# Limitation
+
+* YAML is a superset of JSON, meaning there are features supported in YAML but not in JSON. Most notable is YAML's inline comment feature which does not have an equivalent in JSON. If `Fn::ToJsonString` is used on a YAML template block that contains comments, the comments will be automatically stripped out when converted to JSON. This can potentially be a confusing developer experience if the developer is not aware of these limitations and have different expectations.
+* Unsupported pseudo parameters:
+    * `AWS::NotificationARNs`
+* The object passed to the `Fn::ToJsonString` intrinsic function must be either a `Map` or an `Array`
```
Comment (MalikAtalla-AWS) [inline RFCs/0014-ToJsonString.md:110]: Nitpick: Isn't the term `Object` more common than `Map` when talking about Json? For yaml, both Map and object seem to be used.
```suggestion
* The object passed to the `Fn::ToJsonString` intrinsic function must be either an `Object` or an `Array`. Simple types such as number, boolean or string are not supported.
```

**Review by arthurboghossian — APPROVED** [review]

**Review by MalikAtalla-AWS — APPROVED** [review]

**Pull request merged**